pkg/proxy/endpoints.go

```diff
@@ -396,23 +399,29 @@ func (em EndpointsMap) apply(ect *EndpointChangeTracker, staleEndpoints *[]Servi
 
 	changes := ect.checkoutChanges()
 	for _, change := range changes {
-		em.unmerge(change.previous)
-		em.merge(change.current)
+		em.unmerge(change.previous, ect.endPointMapEntryChanged)
+		em.merge(change.current, ect.endPointMapEntryChanged)
```

Comment:
Is it possible to rework this to call a single function with all the changes here instead? Maybe something like `ect.onEndpointChange(change)`, or `ect.onEndpointChanges(changes)`.

Reply:
I am not sure if i understand. The merge and unmerge are basically operations on the endpointMap.  Can you please give some more context

Reply:
So the merge/unmerge functions now look something like this:
```golang
for svcPortName := range other {
	if changeFunc != nil {
		changeFunc(&svcPortName)
	}
	...
}
```

Depending on the function, `other` represents `change.previous` or `change.current`. If I understand what you want to accomplish here, it's to be notified anytime endpoints change for a unique ServicePortName. My suggestion was to have a single call to `ect.onEndpointChanges(changes)` and then have this loop in the proxier instead. Essentially just limiting how much needs to change in the shared code - 1 function call vs multiple.

Reply:
I wanted to keep the changes in the IPVS and Iptables proxier minimal. The only change is passing nil to these handlers to make this change relevant to WinKernel. The existing pattern is to take the changes in the changetracker and apply it to the map. I would like to keep this as  is unless you see any functional issues with this .

Reply:
I agree with keeping the changes minimal to ipvs and iptables proxiers, I just think we can also minimize the changes to the shared code as well. Instead of passing to and then calling the changeFunc in each merge/unmerge function, that could be called after merge/unmerge are called. So something more like this:

```golang
em.unmerge(change.previous)
em.merge(change.current)
if ect.onEndpointsMapChange != nil {
    ect.onEndpointsMapChange(change.previous, change.current)
}
```

This limits changes to the shared code and gives implementations a way to hook into the details of any endpointMapChange for any reason, not just this use case. I also think it might make sense to change this function name to `onEndpointsMapChange` to match some of the other naming. 

As a sidenote, if I'm understanding the current implementation correctly, it looks like the `endpointMapChange` func is called before the chance is actually applied to the service or endpointsMap. Since the winkernel functions rely on what's inside endpointsMap as they're called, it seems like the order may need to change here.

Reply:
Makes sense. Will do this change